### PR TITLE
[Core] Included files in json

### PR DESCRIPTION
**Pull request by AriadnaCortesDanes**
**📝 Description**
Allows json files to include other json files using the keyword "@include_json" (issue #6815). The code adds a post-process to the original json data, and substitutes "@include_json" entries by the contents of the new json file especified. 

Just to illustrate. If we have a first json: 
`{
"data1" : "hello",
"@include_json" : "foo.json",
"data2" : 2.0
}`

and the file foo.json contains:
`{
"data3" :  3.0,
"data4" : 4.0
}`

the final processed data would be equivalent as if we had:  
`{
"data1" : "hello",
"data2" : 2.0,
"data3" :  3.0,
"data4" : 4.0
}`

Note that the function is recursive in the number of includes.

**🆕 Changelog**
- Added a function (and header) in order to read a json file
- Added a function (and header) in order to post-process the json data
- Added corresponding tests

### CODE REVIEW

File: kratos/sources/kratos_parameters.cpp
Comment (loumalouomega) [inline kratos/sources/kratos_parameters.cpp]: This is not external dependency, it goes in the upper one

File: kratos/sources/kratos_parameters.cpp
Comment (loumalouomega) [inline kratos/sources/kratos_parameters.cpp]: Input is always CammelCase
```suggestion
nlohmann::json Parameters::ReadFile(std::string FileName) {
```

File: kratos/sources/kratos_parameters.cpp
Comment (loumalouomega) [inline kratos/sources/kratos_parameters.cpp]:
```suggestion
void Parameters::SolveIncludes(nlohmann::json& rJson) 
{
```
Style

File: kratos/includes/kratos_parameters.h
Comment (loumalouomega) [inline kratos/includes/kratos_parameters.h]:
```suggestion
```
I assume it is a leftover

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some comments. Implementation simple and elegant

File: kratos/includes/kratos_parameters.h
Comment (loumalouomega) [inline kratos/includes/kratos_parameters.h]:
```suggestion
    nlohmann::json ReadFile(std::string FileName);
```

**Review by loumalouomega — DISMISSED** [review]

**Review by loumalouomega — DISMISSED** [review]

**Comment by pooyan-dadvand** [comment]
I think this needs the approval of the @KratosMultiphysics/technical-committee as a generic core feature

**Comment by loumalouomega** [comment]
New test are failing

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Blocking until @KratosMultiphysics/technical-committee agrees

**Comment by matekelemen** [comment]
Should circular includes be checked or is this something we don't worry about?

**Comment by roigcarlo** [comment]
> Should circular includes be checked or is this something we don't worry about?

We should totally check that if we do not

**Comment by matekelemen** [comment]
> We should totally check that if we do not

I can do that but we should probably get the tests working first. I don't want to commit to this PR without approval so I'll use [another branch](https://github.com/KratosMultiphysics/Kratos/compare/core/included-files-in-json...core/acyclic-json-links?w=1) in the meantime.

> New test are failing

The problem is with the [path to the included json](https://github.com/KratosMultiphysics/Kratos/blob/77f365ce8184ef92b9fc50c0126cbffb9ca82ef2/kratos/tests/cpp_tests/sources/test_kratos_parameters.cpp#L274). Kratos implements tools for changing the working directory during testing for python scripts, so the filesystem can be safely navigated, but there's no such infrastructure for cpp tests.

I vaguely remember someone working on a unified test interface for cpp and python tests, which would be a good place to implement working directory manipulation. Does anyone know anything about the status of that project?

**Comment by roigcarlo** [comment]
> I can do that but we should probably get the tests working first. I don't want to commit to this PR without approval so I'll use [another branch](https://github.com/KratosMultiphysics/Kratos/compare/core/included-files-in-json...core/acyclic-json-links?w=1) in the meantime.

👍 

>  Kratos implements tools for changing the working directory during testing for python scripts, so the filesystem can be safely navigated, but there's no such infrastructure for cpp tests

You should be able to use `Kratos::CurrentWorkingDirectory()` it should be a wrapper for `boost/std::filesystem::current_path` (https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/kratos_filesystem.h) and move from there. Its an external lib now but it will be part of c++17 when we move and we have an abstract interface for it so you will not notice any diference

**Comment by matekelemen** [comment]
Ready for review.

(No idea what's happening on CentOS but I doubt it's related to this PR)

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee Agrees with the changes, but @pooyan-dadvand will make a final review before approving. Gj!

**Comment by ddiezrod** [comment]
any advance on this?

**Comment by loumalouomega** [comment]
> any advance on this?

I think we are waiting for @pooyan-dadvand final review

**Comment by pooyan-dadvand** [comment]
@mpentek Why you store the `adjacency_list`? Why you are not only checking the stack to see if you have the same entry somewhere there?

**Comment by matekelemen** [comment]
> Why you store the adjacency_list? Why you are not only checking the stack to see if you have the same entry somewhere there?

The stack's purpose is to help navigating the JSON structure, not to store ```@include_json``` entries. It stores nested json objects so it can't be used to check for included file names.

Moreover, having duplicate includes doesn't necessarily mean that there's a cycle in the include graph, which is part of the reason why we need a full graph representation of the include structure. That's what ```adjacency_list``` (or ```rAdjacencyMap```) is for.

**Comment by pooyan-dadvand** [comment]
I understand that we can include the same file in different places but that should not appear in the stack twice. So still checking the filename in the stack should be valid. 

In other word, the stack of the calls would be the same as the adjacency map from the root to the given leaf. So I think we can use it to travers the map without really creating map.

**Comment by matekelemen** [comment]
@pooyan-dadvand ok so you don't mean [this stack](https://github.com/KratosMultiphysics/Kratos/blob/c272bd3acd0e46aaf7ed9ce01ab62ce0831abc64/kratos/sources/kratos_parameters.cpp#L345), but rather your suggestion is to replace [the adjacency map](https://github.com/KratosMultiphysics/Kratos/blob/c272bd3acd0e46aaf7ed9ce01ab62ce0831abc64/kratos/sources/kratos_parameters.cpp#L236) with **a** stack that only contains the current sequence of included files? I can see that working.

File: kratos/includes/kratos_filesystem.h
Comment (philbucher) [inline kratos/includes/kratos_filesystem.h]: Can you please make a separate PR for this?
I have some comments

**Comment by pooyan-dadvand** [comment]
@matekelemen thanks for the confirmation. I think stack would be a better choice. And also we can keep track of the include file paths in the same stack of the call. But if you prefer a separate stack I would not oppose it.

@AriadnaCortesDanes would you please proceed with this change?

**Comment by matekelemen** [comment]
> I think stack would be a better choice. And also we can keep track of the include file paths in the same stack of the call.

I'm afraid I don't know how to do that :(

> But if you prefer a separate stack I would not oppose it.

@pooyan-dadvand it's already implemented via cbc87dba43ecc1e185641d6288178fa25fb89d1a.

File: kratos/includes/kratos_parameters.h
Comment (pooyan-dadvand) [inline kratos/includes/kratos_parameters.h]: Do we need this one?

**Review by pooyan-dadvand — DISMISSED** [review]
I think the current stack implementation for the cyclic dependency is ok. So let's go with it!

File: kratos/includes/kratos_parameters.h
Comment (matekelemen) [reply kratos/includes/kratos_parameters.h]: Nope, good catch!

File: kratos/includes/kratos_parameters.h
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
    void SolveIncludes(nlohmann::json& rJson, const std::filesystem::path& rFileName, std::vector<std::string>& rIncludeSequence);
```

File: kratos/includes/kratos_parameters.h
Comment (philbucher) [inline kratos/includes/kratos_parameters.h]:
```suggestion
    nlohmann::json ReadFile(const std::filesystem::path& rFileName);
```

**Review by philbucher — COMMENTED** [review]
lets make use of C++17 since we now finally have it :tada:

**Comment by pooyan-dadvand** [comment]
Seems that windows didn't liked the file system. Maybe a merge from master is needed.

**Comment by loumalouomega** [comment]
> Seems that windows didn't liked the file system. Maybe a merge from master is needed.

C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.33.31629\include\iterator(151,27): error C2039: 'insert': is not a member of 'std::filesystem::path' [D:\a\Kratos\Kratos\build\Custom\kratos\KratosCore.vcxproj] 

Looks like this is the problem

**Comment by matekelemen** [comment]
I'm not quite sure what the problem is but I'll only be able to take a look at it during the weekend. I'll let you know when I fixed it.

**Comment by matekelemen** [comment]
@pooyan-dadvand ok looks like MSVC's satisfied too.

### Details

For anyone who later might fall into the same trap with ```string```s and ```path```s, here's what was wrong:

Unlike any sane toolchain, MSVC's implementation of the standard library stores the data of an ```std::filesystem::path``` in a ```std::wstring``` internally instead of a normal ```std::string```. You might not have noticed this, because implicit conversions between STL types are well implemented and have no ADL-related issues since they're all within the same namespace.

However, if a class outside the ```std``` namespace implements ```operator std::string``` but not ```operator std::wstring``` (such as the ```nlohmann::basic_json``` class), the conversion on MSVC will fail because it'd have to cast through the following chain: ```your_class``` => ```std::string``` => ```std::wstring``` => ```std::filesystem::path```. Solution: explicitly cast your type to something in the ```std``` namespace, and then construct an ```std::filesystem::path``` from that. Thank you Microsoft!

**Comment by philbucher** [comment]
:window: :man_facepalming:

**Review by pooyan-dadvand — APPROVED** [review]
Thank you all!

**Comment by pooyan-dadvand** [comment]
@loumalouomega would you please check if you can unblock the PR?

**Review by loumalouomega — APPROVED** [review]
Approving

**Comment by loumalouomega** [comment]
> @loumalouomega would you please check if you can unblock the PR?

Done

**Pull request merged**
